### PR TITLE
Added support to lock on version in BlobInputStream

### PR DESCRIPTION
**Pull request by gapra-msft**
Resolves https://github.com/Azure/azure-sdk-for-java/issues/16882

Here is a gist of this code.
Added ConsistentReadControl enum - NONE, E_TAG, VERSION_ID - Documentation for each is on the Enum.

Default value - E_TAG
We allow for setting etag and versionid at same time.
If E_TAG or VERSION_ID specified, we will lock on latest if not specified by user.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     * No concurrency control. 'requestConditions.ifMatch' and 'client.versionId' must not be set.
+     */
+    NONE,
+
+    /**
+     * Default value. Concurrency control based on eTag.
+     * If {@link BlobRequestConditions#getIfMatch()} is set, the client will honor this value.
+     * Otherwise, {@link BlobRequestConditions#getIfMatch()} is set to the latest eTag.
+     */
+    E_TAG,
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:25]: `E_TAG` or `ETAG`? In other locations in Storage we already use constants where it doesn't have the `_` such as https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/Constants.java#L221

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 2]
```diff
@@ -261,23 +262,60 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+        BlobAsyncClientBase client = this.client;
+
+        switch (concurrencyControl) {
+            case NONE:
+                if (requestConditions.getIfMatch() != null) {
+                    throw logger.logExceptionAsError(generateNoneException("requestConditions.ifMatch", "E_TAG"));
+                }
+                if (this.client.getVersionId() != null) {
+                    throw logger.logExceptionAsError(generateNoneException("client.versionId", "VERSION_ID"));
+                }
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:286]: Do we want to log an informational or warning message here instead of throwing?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 3]
```diff
@@ -261,23 +262,60 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+                    new UnsupportedOperationException("Concurrency control type not supported."));
         }
 
         return new BlobInputStream(client, range.getOffset(), range.getCount(), chunkSize,
             requestConditions, properties);
     }
 
+    private UnsupportedOperationException generateNoneException(String wrongValue, String toSet) {
+        return new UnsupportedOperationException(String.format("'%s' can not be set when 'concurrencyControl'"
+            + " is set to NONE. Set 'concurrencyControl' to %s.", wrongValue, toSet));
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:316]: Should we make this into two constants as we only have two cases we need to cover and that will eliminate a string format.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 4]
```diff
@@ -261,23 +262,60 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+                } else {
+                    // Target the user specified version by default. If not provided, target the latest version.
+                    if (this.client.getVersionId() == null) {
+                        client = this.client.getVersionClient(versionId);
+                    }
+                }
+                break;
+            default:
+                throw logger.logExceptionAsError(
+                    new UnsupportedOperationException("Concurrency control type not supported."));
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:307]: Should the error types being thrown here be IllegalStateException instead of UnsupportedOperationException? They feel more like bad states and not like operations that can't be perfromed.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 5]
```diff
@@ -261,23 +262,60 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
      */
     public BlobInputStream openInputStream(BlobInputStreamOptions options) {
         options = options == null ? new BlobInputStreamOptions() : options;
+        ConcurrencyControl concurrencyControl = options.getConcurrencyControl() == null ? ConcurrencyControl.E_TAG
+            : options.getConcurrencyControl();
 
         BlobProperties properties = getProperties();
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:268]: I think we should be using the customer's requestConditions for this call and THEN locking on the resulting etag.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 1 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:25]: I think the docs should also specify that we will respect the conditions they pass initially and then we will lock on the etag

Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:25]: And should maybe mention that specifying an etag is mutually exclusive to having a version set.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 6]
```diff
@@ -261,23 +262,60 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+                }
+                break;
+            case VERSION_ID:
+                if (versionId == null) {
+                    throw logger.logExceptionAsError(
+                        new UnsupportedOperationException("Versioning is not supported on this account."));
+                } else {
+                    // Target the user specified version by default. If not provided, target the latest version.
+                    if (this.client.getVersionId() == null) {
+                        client = this.client.getVersionClient(versionId);
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:301]: I think we want to throw here if etag is specified and throw above if version is specified

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.azure.storage.blob.BlobClient;
+
+/**
+ * Defines values for ConcurrencyControl.
+ * {@link #NONE}
+ * {@link #E_TAG}
+ * {@link #VERSION_ID}
+ */
+public enum ConcurrencyControl {
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:14]: I don't love the name concurrency control. I think it's technically correct, but it sounds to me more like it's intended to manage like number of threads or something. And etags are for optimistic concurrency, but I don't think versions fits in that linguistic space. That might just be me. It's more about like ensuring a consistent view of the blob for the whole download. It's also weird because they behave differently. Etag will throw if it changes, but version id will never throw as long as the version is not deleted. 
... I don't have any suggestions here though.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:25]: I can change it to ETAG

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:25]: @rickle-msft - Does this line not indicate that we will respect their conditions?
If {@link BlobRequestConditions#getIfMatch()} is set, the client will honor this value.
   Otherwise, {@link BlobRequestConditions#getIfMatch()} is set to the latest eTag

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:316]: Yeah, this doesn't really need to be a method honestly

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 2 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:286]: I would rather throw. I dont want the user to think they were locking on something but they really werent

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 4 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:307]: I think its more of a Illegal Argument Exception. I can change it to that

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 7 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:14]: ConsistencyControl?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java]: With the addition of a few more error cases (if you take a second look), I think we should just leave this as a format.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 1 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java]: Yea that's fine. I think when I wrote that, I was thinking about the other comment I left about using the customer's request for the getProperties and applying that here. But anyway. Yea this is fine.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 7 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:14]: Yea I think that works better in my mind. Good one.

Comment (kasobol-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:14]: `ConsistentReadControl` ?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 8]
```diff
@@ -261,23 +262,71 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+                // Target the user specified eTag by default. If not provided, target the latest eTag.
+                if (requestConditions.getIfMatch() == null) {
+                    requestConditions.setIfMatch(eTag);
+                }
+                break;
+            case VERSION_ID:
+                if (requestConditions.getIfMatch() != null) {
+                    throw logger.logExceptionAsError(generateControlException("requestConditions.ifMatch",
+                        ConcurrencyControl.VERSION_ID.toString(), ConcurrencyControl.ETAG.toString()));
+                }
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:304]: How about other requestConditions?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 7 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java:14]: I like that as well

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 8 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:304]: I think only etag is checked because we only lock on the etag. The other request conditions are only used in the initial get properties and then will be replaced by the etag (if that's the control type specified). Although I can see how it might be odd to specify LMT and version at the same time....

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:304]: Maybe they just want that extra level of assurance :P

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 7 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java]: Going with Kamil's suggestion

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Defines values for ConcurrencyControl.
+ * {@link #NONE}
+ * {@link #ETAG}
+ * {@link #VERSION_ID}
+ */
+public enum ConsistentReadControl {
+    /**
+     * No concurrency control. 'requestConditions.ifMatch' and 'client.versionId' must not be set.
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:16]: Rewrite these docs

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 10]
```diff
@@ -261,23 +262,71 @@ public final BlobInputStream openInputStream(BlobRange range, BlobRequestConditi
...
+                if (this.client.getVersionId() != null) {
+                    throw logger.logExceptionAsError(generateConsistentReadControlException("client.versionId",
+                        ConsistentReadControl.NONE.toString(), ConsistentReadControl.VERSION_ID.toString()));
+                }
+                break;
+            case ETAG:
+                if (this.client.getVersionId() != null) {
+                    throw logger.logExceptionAsError(generateConsistentReadControlException("client.versionId",
+                        ConsistentReadControl.ETAG.toString(), ConsistentReadControl.VERSION_ID.toString()));
+                }
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:294]: If someone is using both versionid and ifmatch condition this is going to break them.
Maybe we shouldn't default to "ETAG" few lines above but rather apply this switch block only if consistency controll was explicitly set. Otherwise use old logic?

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:294]: I think Rick and I already talked about this - there really is no reason to be using both version and ifMatch conditions (since theyre basically duplicating work).

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 11]
```diff
@@ -173,4 +174,149 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+        try {
+            while ((b = inputStream.read()) != -1) {
+                outputStream.write(b)
+            }
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex)
+        }
+        byte[] randomBytes2 = outputStream.toByteArray()
+        assert randomBytes2 == Arrays.copyOfRange(randomBytes, 1 * Constants.MB, 6 * Constants.MB)
+
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:258]: I think we should also have a case where readControl.ETAG is set, but no request conditions are passed to ensure we will pick the latest etag if none is specified.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 12]
```diff
@@ -173,4 +174,149 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+        then:
+        def outputStream = new ByteArrayOutputStream()
+        def b
+        try {
+            while ((b = inputStream.read()) != -1) {
+                outputStream.write(b)
+            }
+        } catch (IOException ex) {
+            throw new UncheckedIOException(ex)
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:254]: why is this necessary?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 11 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:258]: And some validation that etag and version can't both be set for both the etag and version options.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 13]
```diff
@@ -173,4 +174,149 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+        thrown(UnsupportedOperationException)
+
+        when:
+        blobClient.getVersionClient(properties.getVersionId()).openInputStream(new BlobInputStreamOptions().setBlockSize(1).setConsistentReadControl(ConsistentReadControl.NONE))
+
+        then:
+        thrown(UnsupportedOperationException)
+    }
+
+    @Requires({ liveMode() })
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:230]: Why do these have to be live mode?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java]: Resolved

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 5 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java]: Resolved

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 6 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:311]: Resolved

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java (hunk 7 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConcurrencyControl.java]: Resolved

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java (hunk 9 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java]: Done

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 11 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:258]: I think that is what "Input stream etag lock default" tests for. I can add tests for the etag and version set at same time ones

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 12 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:254]: Not too sure, this was a copy paste from another test that exists. I can just rethrow the exception

Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:254]: Actually we dont even need to catch since groovy doesnt care

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 13 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy]: just fixed this

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 11 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:216]: Added the other tests too

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+public enum ConsistentReadControl {
+    /**
+     * No consistent read control. 'requestConditions.ifMatch' and 'client.versionId' must not be set.
+     */
+    NONE,
+
+    /**
+     * Default value. Consistent read control based on eTag.
+     * If {@link BlobRequestConditions#getIfMatch()} is set, the client will honor this value.
+     * Otherwise, {@link BlobRequestConditions#getIfMatch()} is set to the latest eTag.
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:23]: locking on etag means we will throw if it changes in the meantime

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    ETAG,
+
+    /**
+     * Consistent control based on versionId. Note: Versioning must be supported by the account to use this value.
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:28]: locking on version means you get a consisten read on blob without throwing if it vhanges in the meantime

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+     * No consistent read control. The client will honor user provided {@link BlobRequestConditions#getIfMatch()} and
+     * {@link BlobClient#getVersionId()}.
+     */
+    NONE,
+
+    /**
+     * Default value. Consistent read control based on eTag.
+     * If {@link BlobRequestConditions#getIfMatch()} is set, the client will honor this value.
+     * Otherwise, {@link BlobRequestConditions#getIfMatch()} is set to the latest eTag.
+     * Note: Modification of the base blob will result in an {@code Exception} if eTag is the only form of consistent
```
Comment (kasobol-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:25]: nit: Should we call out specific exception type that will be thrown?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob.models;
+
+import com.azure.storage.blob.BlobClient;
+
+/**
+ * Defines values for ConsistentReadControl.
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:9]: I think we should add a bit more here: "Defines values to indicate what strategy the sdk should use when reading from a blob to ensure the view of the data is consistent and not changed during the read."

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+     * read control being employed.
+     */
+    ETAG,
+
+    /**
+     * Consistent control based on versionId. Note: Versioning must be supported by the account to use this value.
+     * If {@link BlobClient#getVersionId()} is set, the client will honor this value.
+     * Otherwise, {@link BlobClient#getVersionId()} is set to the latest versionId.
+     * Note: Modification of the base blob will not result in an {@code Exception} and allow you to continue reading the
+     * entirety of the appropriate version of the blob determined at the time of opening the {@code InputStream}.
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/ConsistentReadControl.java:35]: "... but it may no longer be the latest data."

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 19]
```diff
@@ -147,30 +148,308 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+        def properties = blobClient.getProperties()
+
+        when:
+        def inputStream = blobClient.openInputStream(new BlobInputStreamOptions().setConsistentReadControl(ConsistentReadControl.ETAG)
+        // User provides eTag to use
+            .setRequestConditions(new BlobRequestConditions().setIfMatch(properties.getETag())))
+        def outputStream = new ByteArrayOutputStream()
+
+        then: "Successful read"
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:231]: I think we should try modifying and check for failure on these after we get success to make sure that it's actually getting set.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 20]
```diff
@@ -147,30 +148,308 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+        def blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        def blobClient = blobContainerClient.getBlobClient(generateBlobName())
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+        def properties = blobClient.getProperties()
+
+        when:
+        // User provides version client
+        def inputStream = blobClient.getVersionClient(properties.getVersionId()).openInputStream(new BlobInputStreamOptions().setConsistentReadControl(ConsistentReadControl.ETAG))
+        def outputStream = new ByteArrayOutputStream()
+        // When a versioned client is used it should still succeed if the blob has been modified
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:280]: Is this true if the specify etag? Or does the etag on the version also get locked so they'll always match?

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 19 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:231]: I see there's an error case below now.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy [hunk 21]
```diff
@@ -147,30 +148,308 @@ class BlockBlobInputOutputStreamTest extends APISpec {
...
+
+    def "IS consistent read control version user provides version"() {
+        setup:
+        int length = Constants.KB
+        byte[] randomBytes = getRandomByteArray(length)
+        def blobContainerClient = versionedBlobServiceClient.createBlobContainer(containerName)
+        def blobClient = blobContainerClient.getBlobClient(generateBlobName())
+        blobClient.upload(new ByteArrayInputStream(randomBytes), length, true)
+        def properties = blobClient.getProperties()
+
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:349]: Can we do another upload after we get the properties to force this version to be different from the latest at the time of constructing the stream?

**Review by rickle-msft — APPROVED** [review]

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy (hunk 20 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobInputOutputStreamTest.groovy:280]: Since the user specifies the version client - we do a getProperties on that version so we get back the etag for that version so it works right. Does that make sense?

If they specify the etag and it is correct - it'll work. If its wrong it will fail as usual and the service will handle that
